src/crimson/net/ProtocolV2.cc

```diff
@@ -528,21 +559,20 @@ seastar::future<> ProtocolV2::client_auth(std::vector<uint32_t> &allowed_methods
   // send_auth_request() logic
   ceph_assert(messenger.get_auth_client());
 
-  bufferlist bl;
-  vector<uint32_t> preferred_modes;
-  int r = messenger.get_auth_client()->get_auth_request(
-      conn.shared_from_this(), auth_meta, &auth_meta->auth_method,
-      &preferred_modes, &bl);
-  if (r < 0) {
-    logger().error("{} get_initial_auth_request returned {}", conn, r);
+  try {
+    auto [auth_method, preferred_modes, bl] =
+      messenger.get_auth_client()->get_auth_request(conn.shared_from_this(), auth_meta);
+    auth_meta->auth_method = auth_method;
+    auto frame = AuthRequestFrame::Encode(auth_method, preferred_modes, bl);
+    return write_frame(frame).then([this] {
+      return handle_auth_reply();
+    });
+  } catch (const std::system_error& e) {
```

Comment:
ditto.

Reply:
`Encode` and `write_frame()` should not be wrapped in the try block because any error throw from it should trigger `fault()`, not `close()` and `reset()`.

Reply:
then i need to `std::tie()` the returned values. after a second thought, i feel that the authentication failures are not std::**system**_error. how about throws the exceptions defined in `crimson/auth/Errors.h` ?

Reply:
@rzarzynski let's leave the `std::system_error` to the outer `catch`, and let the `catch` in this function take care of  `ceph::auth::error`.